proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+     * deprecation is included here.
+     */
+    silenceDeprecations?: (Deprecation | string)[];
+
+    /**
+     * A set of future deprecations to opt into early.
+     *
+     * If a string is provided, it will resolve to the deprecation with that ID,
+     * or error if none exists.
+     *
+     * This will emit a warning if a non-future deprecation is included here.
+     */
+    futureDeprecations?: (Deprecation | string)[];
+  }
+
+  interface Logger {
+    /**
+     * Update the third sub-bullet of bullet two to read:
+     *
+     * If this warning is caused by behavior that used to be allowed but will
+     * be disallowed in the future, set `options.deprecation` to `true` and
+     * set `options.deprecationType` to the relevant `Deprecation`. Otherwise,
+     * set `options.deprecation` to `false` and leave `options.deprecationType`
+     * undefined.
+     */
+    warn?(
+      message: string,
+      options: {
+        deprecation: boolean;
+        deprecationType?: Deprecation;
+        span?: SourceSpan;
+        stack?: string;
+      }
+    ): void;
+  }
+
+  /**
+   * An object containing all of the deprecations recognized by this
+   * implementation as properties, as well as a `find` function to get the
+   * deprecation for a given ID.
+   */
+  export const deprecations: Readonly<
+    _Deprecations & {[key: string]: Deprecation}
+  >;
+}
+
+interface _Deprecations {
+  /** Deprecation for passing a string to `call` instead of `get-function`. */
+  callString: Deprecation<'call-string', 'active'>;
+
+  /** Deprecation for `@elseif`. */
+  elseif: Deprecation<'elseif', 'active'>;
+
+  /** Deprecation for parsing `@-moz-document`. */
+  mozDocument: Deprecation<'moz-document', 'active'>;
+
+  /** Deprecation for importers using relative canonical URLs. */
+  relativeCanonical: Deprecation<'relative-canonical', 'active'>;
+
+  /** Deprecation for declaring new variables with `!global`. */
+  newGlobal: Deprecation<'new-global', 'active'>;
+
+  /**
+   * Deprecation for certain functions in the color module matching the
+   * behavior of their global counterparts for compatibility reasons.
+   */
+  colorModuleCompat: Deprecation<'color-module-compat', 'active'>;
+
+  /** Deprecation for treaing `/` as division. */
+  slashDiv: Deprecation<'slash-div', 'active'>;
+
+  /** Deprecation for leading, trailing, and repeated combinators. */
+  bogusCombinators: Deprecation<'bogus-combinators', 'active'>;
+
+  /** Deprecation for ambiguous `+` and `-` operators. */
+  strictUnary: Deprecation<'strict-unary', 'active'>;
+
+  /** Deprecation for passing invalid units to certain built-in functions. */
+  functionUnits: Deprecation<'function-units', 'active'>;
+
+  /** Deprecation for `@import` rules. */
+  importRules: Deprecation<'import', 'future'>;
+
+  /** Used for deprecations coming from user-authored code. */
+  userAuthored: Deprecation<'user-authored', 'user'>;
```

Comment:
It's a bit of a pain, but it's probably worth explicitly saying here "update the proposal for XYZ to say that it emits deprecations with id QRS" so that the IDs end up colocated with the text that describes when they're emitted.

Reply:
I mean in the spec comments for each of these deprecation types, literally writing "update the proposal for XYZ to say that it emits deprecations with id QRS". (Language proposals are generally intended to specify modifications to existing spec text.)

Reply:
Okay, I added comments for each of the ones that had a matching proposal. `call-string`, `elseif`, `moz-document`, `relative-canonical`, `new-global`, and `color-module-compat` didn't as far as I could find, so I'm not sure if there's something else I should add here.